### PR TITLE
Define __iter__ and __len__ for Moment

### PR DESCRIPTION
**Pull request by kevinsung**

### CODE REVIEW

**Comment by kevinsung** [comment]
@Strilanc Moments are already valid things to pass into `circuit.append`, with your desired behavior.

**Comment by Strilanc** [comment]
Even when they're in a list?

**Comment by kevinsung** [comment]
@Strilanc I've modified `append` to handle a Sequence of Moments.

File: cirq/circuits/circuit.py [hunk 1]
```diff
@@ -1075,15 +1075,24 @@ def batch_insert(self,
...
+                to the circuit
             strategy: How to pick/create the moment to put operations into.
         """
-        self.insert(len(self._moments), moment_or_operation_tree, strategy)
+        if (isinstance(moments_or_operation_tree, Sequence)
+                and len(moments_or_operation_tree) > 0
+                and isinstance(moments_or_operation_tree[0], Moment)):
+            for moment in moments_or_operation_tree:
+                self.insert(len(self._moments), moment, strategy)
+        else:
```
[the remark this answers is not in the document]
Comment (kevinsung) [reply cirq/circuits/circuit.py]: Ah, I somehow didn't realize until now that this change makes Moments satisfy the OP_TREE contract.

[the remark this answers is not in the document]
Comment (kevinsung) [reply cirq/circuits/circuit.py]: ... which seems to necessitate modifying `flatten_op_tree` to yield Moments in addition to Operations. @Strilanc what do you think?

[the remark this answers is not in the document]
Comment (bryano) [reply cirq/circuits/circuit.py]: But then the output of `flatten_op_tree` wouldn't be flat.

[the remark this answers is not in the document]
Comment (kevinsung) [reply cirq/circuits/circuit.py]: True, that doesn't seem good. I guess I could rewrite the insertion logic to not use `flatten_op_tree`.

[the remark this answers is not in the document]
Comment (kevinsung) [reply cirq/circuits/circuit.py]: Or, we could add a `preserve_moments` boolean flag to `flatten_op_tree`.

[the remark this answers is not in the document]
Comment (Strilanc) [reply cirq/circuits/circuit.py]: I would add the `preserve_moments` flag.

File: cirq/circuits/circuit.py [hunk 2]
```diff
@@ -763,16 +763,16 @@ def _can_commute_past(self,
     def insert(
             self,
             index: int,
-            moment_or_operation_tree: Union[Moment, ops.OP_TREE],
+            operations: ops.OP_TREE,
```
Comment (Strilanc) [inline cirq/circuits/circuit.py]: I think the moment or operation tree name is better, because it indicates to the user that moments are handled in a special way.

File: cirq/circuits/moment.py [hunk 3]
```diff
@@ -14,9 +14,11 @@
 
 """A simplified time-slice of operations within a sequenced circuit."""
 
-from typing import Any, Iterable, TypeVar, Callable, Sequence
+from typing import Any, Iterable, TypeVar, Callable, Sequence, TYPE_CHECKING
 
-from cirq import ops
+if TYPE_CHECKING:
+    # pylint: disable=unused-import
+    from cirq import ops
```
Comment (Strilanc) [inline cirq/circuits/moment.py]: Why did you need to move this into a type-checking only block? There shouldn't be any dependence of ops on circuits, and if there is the cycle should be broken there.

File: cirq/ops/op_tree.py [hunk 4]
```diff
@@ -19,6 +19,7 @@
 
 from typing import Any, Iterable, Union, Callable
 
+from cirq.circuits.moment import Moment
```
Comment (Strilanc) [inline cirq/ops/op_tree.py]: Ah. Break cycle here instead of elsewhere.

Comment (kevinsung) [reply cirq/ops/op_tree.py]: I don't see a way to implement the `preserve_moments` boolean flag without using Moment in `isinstance` checks.

**Comment by kevinsung** [comment]
mypy is very unhappy with this change ...

File: cirq/circuits/circuit.py [hunk 5]
```diff
@@ -787,16 +787,17 @@ def _can_commute_past(self,
     def insert(
             self,
             index: int,
-            moment_or_operation_tree: Union[Moment, ops.OP_TREE],
+            moment_or_operation_tree: ops.OP_TREE,
```
Comment (Strilanc) [inline cirq/circuits/circuit.py]: This is not the correct type.

I recommend defining `OP_OR_MOMENT_TREE = Union[Operation, Moment, OP_TREE, Iterable[Any]]` in order to appease mypy.

File: cirq/ops/op_tree.py (hunk 4 above)
Comment (Strilanc) [reply cirq/ops/op_tree.py]: That sounds reasonable.

File: cirq/circuits/circuit.py (hunk 5 above)
Comment (kevinsung) [reply cirq/circuits/circuit.py]: Why isn't it the correct type? Operation and Moment are both examples of OP_TREE.

Comment (Strilanc) [reply cirq/circuits/circuit.py]: That's... Huh. I guess that's right. It just seems really weird to tell the user you're asking for "operations or moments" and then just say op tree.

Comment (kevinsung) [reply cirq/circuits/circuit.py]: That's why I initially renamed to variable to just `operations`. Also, mypy was mainly unhappy about the return type of `flatten_op_tree` being changed from `Iterable[Operation]` to `Iterable[Union[Operations, Moment]]`.

Comment (Strilanc) [reply cirq/circuits/circuit.py]: You would need to define overloads that said OP_TREE mapped to OP_TREE while the more complicated type mapped to the more complicated type.

Comment (kevinsung) [reply cirq/circuits/circuit.py]: @Strilanc do you still want me to define `OP_OR_MOMENT_TREE` or are you okay with just leaving `OP_TREE` and changing the variable name to simply `operations`?

Comment (Strilanc) [reply cirq/circuits/circuit.py]: In my opinion the variable name needs to say moments because that's important documentation that will affect user expectations. And the type name should probably also say moments. So yes, please change it.

Comment (kevinsung) [reply cirq/circuits/circuit.py]: I've renamed the variable to `moment_or_operation_tree` and the type to `Union[Moment, OP_TREE]`.
